File: Data Structures & Algorithms/max-water-container/submission-1.java
class Solution {
    public int maxArea(int[] heights) {
        int N = heights.length;
        int max = 0;

        int left = 0;
        int right = N - 1;

        while (left < right) {
            int height = Math.min(heights[left], heights[right]);

            max = Math.max(height * (right - left), max);

            if (heights[left] > heights[right]) {
                right --;
            } else {
                left ++;
            }
        }
        return max;
    }
}

File: Data Structures & Algorithms/eating-bananas/submission-1.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        left, right = 1, max(piles)
        res = right

        while left <= right:
            speed = (left + right) // 2
            hours = 0
            for p in piles:
                hours += math.ceil(float(p) / speed)
            
            if hours <= h:
                res = min(res, speed)
                right = speed - 1
            else:
                left = speed + 1

        return res